packages/interaction-tracking/src/InteractionTracking.js

```diff
@@ -0,0 +1,341 @@
+  // Work is done by a "thread" which is identified by a unique ID.
+  onWorkStarted: (interactions: Set<Interaction>, threadID: number) => void,
+
+  // A batch of work has completed for a set of interactions.
+  // Work is done by a "thread" which is identified by a unique ID.
+  onWorkStopped: (interactions: Set<Interaction>, threadID: number) => void,
+};
+
+export type InteractionsRef = {
+  current: Set<Interaction>,
+};
+
+export type SubscriberRef = {
+  current: Subscriber | null,
+};
+
+const DEFAULT_THREAD_ID = 0;
+
+// Counters used to generate unique IDs.
+let interactionIDCounter: number = 0;
+let threadIDCounter: number = 0;
+
+// Set of currently tracked interactions.
+// Interactions "stack"–
+// Meaning that newly tracked interactions are appended to the previously active set.
+// When an interaction goes out of scope, the previous set (if any) is restored.
+let interactionsRef: InteractionsRef | null = null;
+
+// Listener(s) to notify when interactions begin and end.
+// Note that subscribers are only supported when enableInteractionTrackingObserver is enabled.
+let subscriberRef: SubscriberRef | null = null;
+
+if (enableInteractionTracking) {
+  interactionsRef = {
+    current: new Set(),
+  };
+  if (enableInteractionTrackingObserver) {
+    subscriberRef = {
+      current: null,
+    };
+  }
+}
+
+// These values are exported for libraries with advanced use cases (i.e. React).
+// They should not typically be accessed directly.
+export {interactionsRef as __interactionsRef, subscriberRef as __subscriberRef};
+
+export function clear(callback: Function): any {
+  if (!enableInteractionTracking) {
+    return callback();
+  }
+
+  const prevInteractions = ((interactionsRef: any): InteractionsRef).current;
+  ((interactionsRef: any): InteractionsRef).current = new Set();
+
+  try {
+    return callback();
+  } finally {
+    ((interactionsRef: any): InteractionsRef).current = prevInteractions;
+  }
+}
+
+export function getCurrent(): Set<Interaction> | null {
+  if (!enableInteractionTracking) {
+    return null;
+  } else {
+    return ((interactionsRef: any): InteractionsRef).current;
+  }
+}
+
+export function getThreadID(): number {
+  return ++threadIDCounter;
+}
+
+export function subscribe(subscriber: Subscriber): void {
+  if (enableInteractionTracking && enableInteractionTrackingObserver) {
+    invariant(
+      ((subscriberRef: any): SubscriberRef).current === null ||
+        ((subscriberRef: any): SubscriberRef).current === subscriber,
+      'Only one interactions subscriber may be registered at a time.',
+    );
+    ((subscriberRef: any): SubscriberRef).current = subscriber;
+  }
+}
+
+export function track(
+  name: string,
+  callback: Function,
+  threadID: number = DEFAULT_THREAD_ID,
+): any {
+  if (!enableInteractionTracking) {
+    return callback();
+  }
+
+  const interaction: Interaction = {
+    id: interactionIDCounter++,
+    name,
+    timestamp: now(),
```

Comment:
Since there is nothing else in here that deals with timestamps, it seems better that this is passed in instead. That way the time mechanism doesn’t have to be built in. Eg we are considering a custom measurer at FB and RN might need something else.

In fact for certain tracking I believe that is required. Eg for a click you want the time at the time the user released. Not towards the end of the processing of the click event. The mouseup event could’ve spent time too. For requestAnimationFrame you want the time stamp passed into the callback (beginning of the frame), not the current time.

Additionally we don’t even know which scale is appropriate to use. Date.now and performance.now use different time scales and whatever they’re going to be compared to (eg) commit time needs to line up.

If we’re going to automatically track time here, then we should also pass time to the complete callbacks so that the external system gets consistent time. But I see no reason for that.

We can just have the user of the track api pass the current time in as an argument. That makes this package a bit lighter too.

Reply:
Hmm...that would make the API feel clunkier to me ☹️but if you think it's important/necessary, I trust your judgement.

> That makes this package a bit lighter too.

By such a tiny amount that it isn't really important 😁